server/schedule/labeler/labeler.go

```diff
@@ -31,23 +33,73 @@ import (
 type RegionLabeler struct {
 	storage endpoint.RuleStorage
 	sync.RWMutex
-	labelRules map[string]*LabelRule
-	rangeList  rangelist.List // sorted LabelRules of the type `KeyRange`
+	labelRules        map[string]*LabelRule
+	rangeList         rangelist.List // sorted LabelRules of the type `KeyRange`
+	ctx               context.Context
+	earlistExpireTime time.Time
 }
 
 // NewRegionLabeler creates a Labeler instance.
-func NewRegionLabeler(storage endpoint.RuleStorage) (*RegionLabeler, error) {
+func NewRegionLabeler(ctx context.Context, storage endpoint.RuleStorage, gcInterval time.Duration) (*RegionLabeler, error) {
 	l := &RegionLabeler{
 		storage:    storage,
 		labelRules: make(map[string]*LabelRule),
+		ctx:        ctx,
 	}
 
 	if err := l.loadRules(); err != nil {
 		return nil, err
 	}
+	go l.doGC(gcInterval)
 	return l, nil
 }
 
+func (l *RegionLabeler) doGC(gcInterval time.Duration) {
+	ticker := time.NewTicker(gcInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			l.checkAndClearExpiredLabels()
+			log.Debug("RegionLabeler GC")
+		case <-l.ctx.Done():
+			log.Info("RegionLabeler GC stopped")
+			return
+		}
+	}
+}
+
+func (l *RegionLabeler) checkAndClearExpiredLabels() {
+	now := time.Now()
+	l.Lock()
+	defer l.Unlock()
+
+	if l.earlistExpireTime.After(now) {
+		return
+	}
+	var err error
+	deleted := false
+
+	for key, rule := range l.labelRules {
+		if !rule.checkAndRemoveExpireLabels(now) {
+			continue
+		}
+		if len(rule.Labels) == 0 {
+			err = l.storage.DeleteRegionRule(key)
```

Comment:
It will just log the error info.  However, if there is something wrong with the storage last for long, the restart of the server would be long with too much expired items.